PETWorks/profitability.py

```diff
@@ -0,0 +1,238 @@
+
+
+def _setDataHierarchies(
+    data: Data, hierarchies: dict[str, list[list[str]]], attributeTypes: dict
+) -> None:
+    for attributeName, hierarchy in hierarchies.items():
+        data.getDefinition().setAttributeType(attributeName, hierarchy)
+        attributeType = attributeTypes.get(attributeName)
+
+        if attributeType == IDENTIFIER:
+            data.getDefinition().setAttributeType(
+                attributeName, AttributeType.IDENTIFYING_ATTRIBUTE
+            )
+
+        if attributeType == INSENSITIVE_ATTRIBUTE:
+            data.getDefinition().setAttributeType(
+                attributeName, AttributeType.INSENSITIVE_ATTRIBUTE
+            )
+
+
+def _getQiIndices(dataHandle: str) -> list[int]:
+    qiNames = dataHandle.getDefinition().getQuasiIdentifyingAttributes()
+    qiIndices = []
+    for qi in qiNames:
+        qiIndices.append(dataHandle.getColumnIndexOf(qi))
+    return qiIndices
+
+
+def _findAnonymousLevel(
+        hierarchy: list[list[str]], value: str
+) -> int:
+    for i in range(len(hierarchy)):
+        for j in range(len(hierarchy[i])):
+            if hierarchy[i][j] == value:
+                return j
+    return -1
+
+
+def _getAnonymousLevels(
+    anonymizedSubset: Data,
+    hierarchies: dict[str, list[list[str]]]
+) -> list[int]:
+
+    subsetDataFrame = _getDataFrame(anonymizedSubset.getHandle())
+    subsetRowNum, subsetColNum = subsetDataFrame.shape
+
+    qiIndices = _getQiIndices(anonymizedSubset.getHandle())
+
+    sampleRowIndex = -1
+    allSuppressed = False
+    for i in range(subsetRowNum):
+        for j in range(subsetColNum):
+            if not (j in qiIndices):
+                continue
+
+            if subsetDataFrame.iloc[i][j] != "*":
+                sampleRowIndex = i
+                break
+
+        if sampleRowIndex != -1:
+            break
+
+        allSuppressed = i == subsetRowNum - 1
+
+    anonymousLevels = []
+    for index in range(subsetColNum):
+        if not (index in qiIndices):
+            continue
+
+        value = subsetDataFrame.iloc[sampleRowIndex][index]
+        attributeName = subsetDataFrame.columns[index]
+        hierarchy = hierarchies[attributeName].getHierarchy()
+
+        if allSuppressed:
+            anonymousLevels.append(len(hierarchy[index])-1)
+            continue
+
+        anonymousLevels.append(_findAnonymousLevel(hierarchy, value))
+
+    return anonymousLevels
+
+
+def _getDataFrame(dataHandle: str) -> pd.DataFrame:
+    rowNum = dataHandle.getNumRows()
+    colNum = dataHandle.getNumColumns()
+
+    colNames = []
+    data = []
+    for i in range(rowNum):
+        row = []
+        for j in range(colNum):
+            row.append(dataHandle.getValue(i, j))
+
+            if i == 0:
+                colNames.append(dataHandle.getAttributeName(j))
+
+        data.append(row)
+
+    return pd.DataFrame(data, columns=colNames)
+
+
+def _getSubsetIndices(
+    table: str, subset: str,
```

Comment:
Please insert a line break after `table: str,` and make the argument `subset: str,` easy to recognize. Beside, I suggest using tool `black` to format the code rather than doing it manually.